File: ios/Biinge/Models/Stats.swift
import Foundation

/// Calendar period the statistics screen aggregates watched activity over
enum StatsPeriod: String, Sendable, CaseIterable {
    case week, month, year, all

    /// Chip label
    var title: String {
        switch self {
        case .week: return "Week"
        case .month: return "Month"
        case .year: return "Year"
        case .all: return "All"
        }
    }

    /// Calendar unit the API buckets this period's activity by
    var unit: Calendar.Component {
        switch self {
        case .week, .month: return .day
        case .year: return .month
        case .all: return .year
        }
    }

    /// Label for the want bar, which counts additions rather than the whole list on a bounded period
    var wantLabel: String {
        self == .all ? "Want" : "Added"
    }

    /// Trailing phrase naming the period in captions
    var phrase: String {
        switch self {
        case .week: return "this week"
        case .month: return "this month"
        case .year: return "this year"
        case .all: return "all time"
        }
    }
}

struct AccountStats: Decodable, Sendable {
    let movies: Movies
    let series: Series
    let episodes: Episodes
    // optional so a stats payload without activity or games (older API) still decodes
    let activity: [Bucket]?
    let games: Games?

    /// want is the whole list all-time, and what was added to it inside a bounded period
    struct Movies: Decodable, Sendable {
        let want: Int
        let watched: Int
        let minutes: Int
    }

    /// watched is finished shows all-time, and shows with an episode watched inside a bounded period
    struct Series: Decodable, Sendable {
        let want: Int
        // a show is being watched now, so no bounded period carries this
        let watching: Int?
        let watched: Int
    }

    struct Episodes: Decodable, Sendable {
        let watched: Int
        let minutes: Int
    }

    /// played is games finished all-time, and games finished inside a bounded period.
    /// minutes is IGDB's time to beat those games, not time the user actually played
    struct Games: Decodable, Sendable {
        let want: Int
        // a game is being played now, so no bounded period carries this
        let playing: Int?
        let played: Int
        let minutes: Int
    }

    /// Watched runtime for one activity bucket, split by media kind
    struct Bucket: Decodable, Sendable, Identifiable {
        let date: String
        let movieMinutes: Int
        let tvMinutes: Int

        var id: String { date }
        var totalMinutes: Int { movieMinutes + tvMinutes }

        /// Start of the bucket parsed from the "yyyy-MM-dd" key
        var start: Date {
            let parts = date.split(separator: "-")
            guard parts.count == 3,
                  let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]) else {
                return .distantPast
            }
            var components = DateComponents()
            components.year = year
            components.month = month
            components.day = day
            return Calendar(identifier: .gregorian).date(from: components) ?? .distantPast
        }
    }
}

File: api/internal/app/models/stats.go
package models

import "time"

// StatsPeriod is the calendar period watched statistics are aggregated over
type StatsPeriod string

const (
	StatsPeriodWeek  StatsPeriod = "week"
	StatsPeriodMonth StatsPeriod = "month"
	StatsPeriodYear  StatsPeriod = "year"
	StatsPeriodAll   StatsPeriod = "all"
)

// StatsPeriods lists every period a user's statistics can be cached under
var StatsPeriods = []StatsPeriod{StatsPeriodWeek, StatsPeriodMonth, StatsPeriodYear, StatsPeriodAll}

// String returns the period's wire value
func (p StatsPeriod) String() string {
	return string(p)
}

// Stats is a user's aggregate watch statistics, every count scoped to the period
type Stats struct {
	// Minutes are summed runtimes. Want is the whole list all-time and what was added to it
	// inside a bounded period; SeriesWatched is shows finished all-time and shows with a
	// watched episode inside a bounded one. SeriesWatching and GamesPlaying are nil on a
	// bounded period, since a show or game is in progress now rather than during some past
	// window. GamesMinutes is IGDB's time to beat, not time actually played
	Period          StatsPeriod
	MoviesWant      uint64
	MoviesWatched   uint64
	MoviesMinutes   uint64
	SeriesWant      uint64
	SeriesWatching  *uint64
	SeriesWatched   uint64
	EpisodesWatched uint64
	EpisodesMinutes uint64
	GamesWant       uint64
	GamesPlaying    *uint64
	GamesPlayed     uint64
	GamesMinutes    uint64
	Activity        []WatchBucket
}

// WatchBucket is watched runtime for one activity bucket, split by media kind
type WatchBucket struct {
	Date         time.Time
	MovieMinutes uint64
	TvMinutes    uint64
}

// NewStatsPeriod maps a raw request value to a supported period, defaulting to all
func NewStatsPeriod(value string) StatsPeriod {
	switch period := StatsPeriod(value); period {
	case StatsPeriodWeek, StatsPeriodMonth, StatsPeriodYear:
		return period
	default:
		return StatsPeriodAll
	}
}
